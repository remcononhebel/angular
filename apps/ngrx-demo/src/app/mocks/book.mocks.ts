export const mockBooks = [
  {
    id: 'mock id 1',
    volumeInfo: { authors: ['mock authors 1'], title: 'mock title 1' },
  },
  {
    id: 'mock id 2',
    volumeInfo: { authors: ['mock authors 2'], title: 'mock title 2' },
  },
];
