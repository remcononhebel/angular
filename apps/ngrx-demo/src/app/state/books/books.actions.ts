import { createActionGroup, emptyProps, props } from '@ngrx/store';
import { Book } from '../../model/book.model';

export const BooksPageActions = createActionGroup({
  source: 'Books Page',
  events: {
    'Load Book List': emptyProps(),
  },
});

export const BooksApiActions = createActionGroup({
  source: 'Books API',
  events: {
    'Load Book List Successful': props<{ books: Book[] }>(),
    'Load Book List Failed': props<{ error: string }>(),
  },
});
