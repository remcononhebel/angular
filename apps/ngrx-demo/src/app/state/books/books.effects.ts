import { Injectable } from "@angular/core";
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { catchError, exhaustMap, map, of } from "rxjs";
import { BooksApiService } from "../../service/books-api.service";

@Injectable()
export class BooksEffects {
    loadBooks$ = createEffect(() => this.actions$.pipe(
        ofType('[Books Page] Load Book List'),
        exhaustMap(() => this.booksApiService.getBooks()
            .pipe(
                map(books => ({ type: '[Books API] Load Book List Successful', books })),
                catchError((error) => of({type: '[Books API] Load Book List Failed', error: 'Failed to load books: ' + error.message}))
            )
        )
    ))


    constructor(private readonly actions$: Actions, private readonly booksApiService: BooksApiService) { }
}