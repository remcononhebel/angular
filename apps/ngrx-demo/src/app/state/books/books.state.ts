import { Book } from '../../model/book.model';

export interface BooksState {
  books: Book[];
  error: string | null;
}
