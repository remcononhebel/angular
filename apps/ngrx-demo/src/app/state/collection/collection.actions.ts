import { createActionGroup, emptyProps, props } from '@ngrx/store';

export const CollectionActions = createActionGroup({
  source: 'Collection',
  events: {
    'Add Book': props<{ bookId: string }>(),
    'Remove Book': props<{ bookId: string }>(),
    'Select books': props<{ bookIds: string[] }>(),
    'Deselect all books': emptyProps()
  },
});

