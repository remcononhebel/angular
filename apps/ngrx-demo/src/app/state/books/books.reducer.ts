import { createReducer, on } from '@ngrx/store';
import { BooksApiActions } from './books.actions';
import { BooksState } from './books.state';

export const initialState: BooksState = {
  books: [],
};

export const booksReducer = createReducer(
  initialState,
  on(BooksApiActions.retrievedBookList, (state, { books }) => {
    return {
      ...state,
      books,
    };
  })
);
