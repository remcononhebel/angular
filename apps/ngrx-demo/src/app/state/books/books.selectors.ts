import { createFeatureSelector, createSelector } from '@ngrx/store';
import { BooksState } from './books.state';
import { Book } from '../../model/book.model';

export const selectBooksState = createFeatureSelector<BooksState>('books');

export const selectAvailableBooks = createSelector(
  selectBooksState,
  (booksState: BooksState) => {
    return booksState.books;
  }
);

export const selectNumberOfAvailableBooks = createSelector(
  selectAvailableBooks,
  (books: Book[]) => books.length
);
