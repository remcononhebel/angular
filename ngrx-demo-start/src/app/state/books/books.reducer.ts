import { createReducer } from '@ngrx/store';
import { BooksState } from './books.state';

export const initialState: BooksState = {
  books: [
    {
      id: '1',
      volumeInfo: {
        authors: ['Author 1', 'Co-author 1'],
        title: 'Book 1',
      },
    },
    {
      id: '2',
      volumeInfo: {
        authors: ['Author 2', 'Co-author 2'],
        title: 'Book 2',
      },
    },
  ],
};

export const booksReducer = createReducer(initialState);
