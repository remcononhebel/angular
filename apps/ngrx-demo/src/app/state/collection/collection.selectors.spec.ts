import { mockBooks } from '../../mocks/book.mocks';
import { BooksState } from '../books/books.state';
import { selectBookCollection, selectNumberOfBooksInCollection } from './collection.selectors';
import { CollectionState } from './collection.state';

describe('CollectionSelectors', () => {
  const booksState: BooksState = {
    books: [...mockBooks],
    error: null
  };

  const collectionStateNoBooks: CollectionState = {
    selectedBooks: [],
  };

  const collectionStateOneBook: CollectionState = {
    selectedBooks: [mockBooks[0].id],
  };
  const collectionStateAllBooks: CollectionState = {
    selectedBooks: mockBooks.map(({ id }) => id),
  };

  describe('selectBookCollection', () => {
    it('should return empty array when no books are selected', () => {
      const collectionState: CollectionState = {
        ...collectionStateNoBooks,
      };
      expect(selectBookCollection.projector(booksState, collectionState)).toMatchSnapshot();
    });

    it('should return one book when one book is selected', () => {
      const collectionState: CollectionState = {
        ...collectionStateOneBook,
      };
      expect(selectBookCollection.projector(booksState, collectionState)).toMatchSnapshot();
    });

    it('should return one book when one book is selected', () => {
      const collectionState: CollectionState = {
        ...collectionStateAllBooks,
      };
      expect(selectBookCollection.projector(booksState, collectionState)).toMatchSnapshot();
    });
  });

  describe('selectNumberOfBooksInCollection', () => {
    it('should say that there are no books in the collection if the collection is empty', () => {
      expect(selectNumberOfBooksInCollection.projector(2, [])).toMatchSnapshot();
    });

    it('should say that there is one out of three books in the collection', () => {
      expect(selectNumberOfBooksInCollection.projector(2, [mockBooks[0]])).toMatchSnapshot();
    });

    it('should say that all books are selected', () => {
      expect(selectNumberOfBooksInCollection.projector(2, [...mockBooks])).toMatchSnapshot();
    });
  });
});
