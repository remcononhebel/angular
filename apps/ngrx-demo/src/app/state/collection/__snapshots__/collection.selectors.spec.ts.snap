// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionSelectors selectBookCollection should return empty array when no books are selected 1`] = `[]`;

exports[`CollectionSelectors selectBookCollection should return one book when one book is selected 1`] = `
[
  {
    "id": "mock id 1",
    "volumeInfo": {
      "authors": [
        "mock authors 1",
      ],
      "title": "mock title 1",
    },
  },
]
`;

exports[`CollectionSelectors selectBookCollection should return one book when one book is selected 2`] = `
[
  {
    "id": "mock id 1",
    "volumeInfo": {
      "authors": [
        "mock authors 1",
      ],
      "title": "mock title 1",
    },
  },
  {
    "id": "mock id 2",
    "volumeInfo": {
      "authors": [
        "mock authors 2",
      ],
      "title": "mock title 2",
    },
  },
]
`;

exports[`CollectionSelectors selectNumberOfBooksInCollection should say that all books are selected 1`] = `"All available books selected"`;

exports[`CollectionSelectors selectNumberOfBooksInCollection should say that there are no books in the collection if the collection is empty 1`] = `"No books selected"`;

exports[`CollectionSelectors selectNumberOfBooksInCollection should say that there is one out of three books in the collection 1`] = `"1 / 2 books selected"`;
